# Pandas for Exploratory Data Analysis I 

Pandas is the most prominent Python library for exploratory data analysis (EDA). The functions Pandas supports are integral to understanding, formatting, and preparing our data. Formally, we use Pandas to investigate, wrangle, munge, and clean our data. Pandas is the Swiss Army Knife of data manipulation!

We'll have two coding-heavy sessions on Pandas. In this one, we'll use Pandas to:

* Read in a dataset
* Investigate a dataset's integrity
* Filter, sort, and manipulate a DataFrame's series

## About the Dataset: Adventure Works Cycles

<img src="assets/adventure_works_logo.png" align="right">

For today's Pandas exercises, we will be using a dataset developed by Microsoft for training purposes, for a fictional company known as [Adventure Works Cycles](https://github.com/Microsoft/sql-server-samples/releases/tag/adventureworks) (AWC), a multinational manufacturer and seller of bicycles and accessories.

A full data dictionary for the Production.Product table is reproduced below, which is a description of the fields (columns) in the table (It is also in the class notes, or [at this link](https://www.sqldatadictionary.com/AdventureWorks2014/)):

| Field | Description |
| :-- | :-- |
| **ProductID** | Primary key for Product records
| **Name** | Name of the product
| **ProductNumber** | Unique product identification number
| **MakeFlag** | 0 = Product is purchased, 1 = Product is manufactured in-house.
| **FinishedGoodsFlag** | 0 = Product is not a salable item. 1 = Product is salable
| **Color** | Product color
| **SafetyStockLevel** | Minimum inventory quantity
| **ReorderPoint** | Inventory level that triggers a purchase order or work order
| **StandardCost** | Standard cost of the product
| **ListPrice** | Selling price
| **Size** | Product size
| **SizeUnitMeasureCode** | Unit of measure for the Size column
| **WeightUnitMeasureCode** | Unit of measure for the Weight column
| **DaysToManufacture** | Number of days required to manufacture the product
| **ProductLine** | R = Road, M = Mountain, T = Touring, S = Standard
| **Class** | H = High, M = Medium, L = Low
| **Style** | W = Womens, M = Mens, U = Universal
| **ProductSubcategoryID** | Product is a member of this product subcategory. Foreign key to ProductSubCategory.ProductSubCategoryID
| **ProductModelID** | Product is a member of this product model. Foreign key to ProductModel.ProductModelID
| **SellStartDate** | Date the product was available for sale
| **SellEndDate** | Date the product was no longer available for sale
| **DiscontinuedDate** | Date the product was discontinued
| **rowguid** | ROWGUIDCOL number uniquely identifying the record. Used to support a merge replication sample
| **ModifiedDate** | Date and time the record was last updated

## Importing Pandas

To [import a library](https://docs.python.org/3/reference/import.html), we write `import` and the library name. For Pandas, is it common to nickname the library `pd` so that when we reference a function from the Pandas library, we only have to write `pd` -- and not `pandas`, just a bit shorter. (This is called [namespacing](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces).)

In [1]:
import pandas as pd

We can see the details about the imported package by referencing its private class variables:

In [2]:
print(f'I am using {pd.__name__} Version: {pd.__version__}.')
print(f'It is installed at: {pd.__path__}')

I am using pandas Version: 0.25.1.
It is installed at: ['/Users/rebeccapeterson/.pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages/pandas']


---
# The Basics

## Reading in Data

Pandas dramatically simplifies the process of reading in data. By "reading in data," we mean loading a file into our machine's memory.

When you have a CSV, for example, and then you double-click to open it in Microsoft Excel, the open file is "read into memory." You can now manipulate the spreadsheet.

When we read data into memory in Python, we are creating object -- specifically, an instance of a `DataFrame`. We will soon explore this object. *(**Note**: as an aside, when we have a file that is greater than the size of our computer's memory, this is approaching "Big Data.")*

Because we are working with a CSV, we will use the [read CSV](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) method.

A [delimiter](https://en.wikipedia.org/wiki/Delimiter-separated_values) is a character that separates fields (columns) in the imported file. Just because a file says `.csv` (recall: Comma Separated Values) does not necessarily mean that a comma is used as the delimiter. In this case, we have a tab character as the delimiter for our columns, so we will be using `sep='\t'` to tell pandas to 'cut' the columns every time it sees a [tab character in the file](http://vim.wikia.com/wiki/Showing_the_ASCII_value_of_the_current_character).

Here's how you read the dataset as a DataFrame into a variable named `products`:

In [6]:
products = pd.read_csv('data/Production.Product.csv', sep='\t')

Use `.head()` to show the head of this dataset:

In [7]:
products.head()

ProductID                   Name ProductNumber  MakeFlag  \
0          1        Adjustable Race       AR-5381         0   
1          2           Bearing Ball       BA-8327         0   
2          3        BB Ball Bearing       BE-2349         1   
3          4  Headset Ball Bearings       BE-2908         0   
4        316                  Blade       BL-2036         1   

   FinishedGoodsFlag Color  SafetyStockLevel  ReorderPoint  StandardCost  \
0                  0   NaN              1000           750           0.0   
1                  0   NaN              1000           750           0.0   
2                  0   NaN               800           600           0.0   
3                  0   NaN               800           600           0.0   
4                  0   NaN               800           600           0.0   

   ListPrice  ... ProductLine Class Style  ProductSubcategoryID  \
0        0.0  ...         NaN   NaN   NaN                   NaN   
1        0.0  ...         NaN   NaN   NaN                   NaN   
2        0.0  ...         NaN   NaN   NaN                   NaN   
3        0.0  ...         NaN   NaN   NaN                   NaN   
4        0.0  ...         NaN   NaN   NaN                   NaN   

   ProductModelID        SellStartDate SellEndDate DiscontinuedDate  \
0             NaN  2008-04-30 00:00:00         NaN              NaN   
1             NaN  2008-04-30 00:00:00         NaN              NaN   
2             NaN  2008-04-30 00:00:00         NaN              NaN   
3             NaN  2008-04-30 00:00:00         NaN              NaN   
4             NaN  2008-04-30 00:00:00         NaN              NaN   

                                  rowguid                   ModifiedDate  
0  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}  2014-02-08 10:01:36.827000000  
1  {58AE3C20-4F3A-4749-A7D4-D568806CC537}  2014-02-08 10:01:36.827000000  
2  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}  2014-02-08 10:01:36.827000000  
3  {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}  2014-02-08 10:01:36.827000000  
4  {E73E9750-603B-4131-89F5-3DD15ED5FF80}  2014-02-08 10:01:36.827000000  

[5 rows x 25 columns]

### Documentation Pause

How did we know how to use `pd.read_csv`? Let's take a look at the [documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). Note the first parameter required (`filepath_or_buffer`) as well as the `sep` parameter.

Right now we don't care too much about the other parameters, suffice to know they exist and you may need them someday!

### Back to `products`

We have just created a data structure called a `DataFrame`. See?

In [8]:
type(products)

pandas.core.frame.DataFrame

## Inspecting our DataFrame

We'll now perform basic operations on the DataFrame.

Here's how to output the first 3 rows:

In [9]:
products.head(3)

ProductID             Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
0          1  Adjustable Race       AR-5381         0                  0   
1          2     Bearing Ball       BA-8327         0                  0   
2          3  BB Ball Bearing       BE-2349         1                  0   

  Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  ...  \
0   NaN              1000           750           0.0        0.0  ...   
1   NaN              1000           750           0.0        0.0  ...   
2   NaN               800           600           0.0        0.0  ...   

  ProductLine Class Style  ProductSubcategoryID  ProductModelID  \
0         NaN   NaN   NaN                   NaN             NaN   
1         NaN   NaN   NaN                   NaN             NaN   
2         NaN   NaN   NaN                   NaN             NaN   

         SellStartDate SellEndDate DiscontinuedDate  \
0  2008-04-30 00:00:00         NaN              NaN   
1  2008-04-30 00:00:00         NaN              NaN   
2  2008-04-30 00:00:00         NaN              NaN   

                                  rowguid                   ModifiedDate  
0  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}  2014-02-08 10:01:36.827000000  
1  {58AE3C20-4F3A-4749-A7D4-D568806CC537}  2014-02-08 10:01:36.827000000  
2  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}  2014-02-08 10:01:36.827000000  

[3 rows x 25 columns]

Notice that `.head()` is a method, so it takes arguments.

* What do you think changes if we pass a different number `head()` argument?
* How would we print the last 5 rows?

In [10]:
products.tail(3)

ProductID                Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
501        997  Road-750 Black, 44    BK-R19B-44         1                  1   
502        998  Road-750 Black, 48    BK-R19B-48         1                  1   
503        999  Road-750 Black, 52    BK-R19B-52         1                  1   

     Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  ...  \
501  Black               100            75      343.6496     539.99  ...   
502  Black               100            75      343.6496     539.99  ...   
503  Black               100            75      343.6496     539.99  ...   

    ProductLine Class Style  ProductSubcategoryID  ProductModelID  \
501          R     L     U                    2.0            31.0   
502          R     L     U                    2.0            31.0   
503          R     L     U                    2.0            31.0   

           SellStartDate SellEndDate DiscontinuedDate  \
501  2013-05-30 00:00:00         NaN              NaN   
502  2013-05-30 00:00:00         NaN              NaN   
503  2013-05-30 00:00:00         NaN              NaN   

                                    rowguid                   ModifiedDate  
501  {44CE4802-409F-43AB-9B27-CA53421805BE}  2014-02-08 10:01:36.827000000  
502  {3DE9A212-1D49-40B6-B10A-F564D981DBDE}  2014-02-08 10:01:36.827000000  
503  {AE638923-2B67-4679-B90E-ABBAB17DCA31}  2014-02-08 10:01:36.827000000  

[3 rows x 25 columns]

Use `shape` to identify the "shape" of the data (rows by columns)

In [11]:
products.shape

(504, 25)

Here we have 504 rows, and 25 columns. This is a tuple, so we can extract the parts using *list indexing*:

In [12]:
print('# Rows is', products.shape[0])
print('# Columns is', products.shape[1])

# Rows is 504
# Columns is 25


## Column Headers and Datatypes

Here's how to get the columns of the table:

In [13]:
products.columns

Index(['ProductID', 'Name', 'ProductNumber', 'MakeFlag', 'FinishedGoodsFlag',
       'Color', 'SafetyStockLevel', 'ReorderPoint', 'StandardCost',
       'ListPrice', 'Size', 'SizeUnitMeasureCode', 'WeightUnitMeasureCode',
       'Weight', 'DaysToManufacture', 'ProductLine', 'Class', 'Style',
       'ProductSubcategoryID', 'ProductModelID', 'SellStartDate',
       'SellEndDate', 'DiscontinuedDate', 'rowguid', 'ModifiedDate'],
      dtype='object')

Let's examine the datatypes of the columns.

Note:

* These datatypes were automatically inferred by pandas upon import!
* Strings are considered `object`s by Pandas

In [14]:
products.dtypes

ProductID                  int64
Name                      object
ProductNumber             object
MakeFlag                   int64
FinishedGoodsFlag          int64
Color                     object
SafetyStockLevel           int64
ReorderPoint               int64
StandardCost             float64
ListPrice                float64
Size                      object
SizeUnitMeasureCode       object
WeightUnitMeasureCode     object
Weight                   float64
DaysToManufacture          int64
ProductLine               object
Class                     object
Style                     object
ProductSubcategoryID     float64
ProductModelID           float64
SellStartDate             object
SellEndDate               object
DiscontinuedDate         float64
rowguid                   object
ModifiedDate              object
dtype: object

Let's convert `products.dtypes` into a DataFrame for easier viewing:

In [16]:
pd.DataFrame(products.dtypes, columns=['DataTypes'])

DataTypes
ProductID                 int64
Name                     object
ProductNumber            object
MakeFlag                  int64
FinishedGoodsFlag         int64
Color                    object
SafetyStockLevel          int64
ReorderPoint              int64
StandardCost            float64
ListPrice               float64
Size                     object
SizeUnitMeasureCode      object
WeightUnitMeasureCode    object
Weight                  float64
DaysToManufacture         int64
ProductLine              object
Class                    object
Style                    object
ProductSubcategoryID    float64
ProductModelID          float64
SellStartDate            object
SellEndDate              object
DiscontinuedDate        float64
rowguid                  object
ModifiedDate             object

Why do datatypes matter? What operations could we perform on some datatypes that we could not on others? Note the importance of this in checking dataset integrity.

---

# Selecting Column(s)

We can select columns in two ways. Either we treat the column as a property of the DataFrame:

In [17]:
products.Name

0            Adjustable Race
1               Bearing Ball
2            BB Ball Bearing
3      Headset Ball Bearings
4                      Blade
               ...          
499        ML Bottom Bracket
500        HL Bottom Bracket
501       Road-750 Black, 44
502       Road-750 Black, 48
503       Road-750 Black, 52
Name: Name, Length: 504, dtype: object

Or we index the DataFrame on the column name:

In [18]:
products['Name']

0            Adjustable Race
1               Bearing Ball
2            BB Ball Bearing
3      Headset Ball Bearings
4                      Blade
               ...          
499        ML Bottom Bracket
500        HL Bottom Bracket
501       Road-750 Black, 44
502       Road-750 Black, 48
503       Road-750 Black, 52
Name: Name, Length: 504, dtype: object

What happens if we use this funny syntax?

In [19]:
products[['Name']]

Name
0          Adjustable Race
1             Bearing Ball
2          BB Ball Bearing
3    Headset Ball Bearings
4                    Blade
..                     ...
499      ML Bottom Bracket
500      HL Bottom Bracket
501     Road-750 Black, 44
502     Road-750 Black, 48
503     Road-750 Black, 52

[504 rows x 1 columns]

In [20]:
print('If I use SINGLE brackets, pandas returns a', 
      type(products['Name']),
      '\nIf I use DOUBLE brackets, pandas returns a',
      type(products[['Name']]))

If I use SINGLE brackets, pandas returns a <class 'pandas.core.series.Series'> 
If I use DOUBLE brackets, pandas returns a <class 'pandas.core.frame.DataFrame'>


Again, if we select the `Name` column only, it is returned as a `Series` object:

In [21]:
products['Name'].head(3)

0    Adjustable Race
1       Bearing Ball
2    BB Ball Bearing
Name: Name, dtype: object

The double brackets really just means a list of columns!

And if we select by a list of columns, it is returned as a DataFrame object:

In [22]:
products[['Name']].head(3)

Name
0  Adjustable Race
1     Bearing Ball
2  BB Ball Bearing

Now that we understand that we're just selecting columns using a list, we can select multiple columns:

In [23]:
products[['Name', 'ProductNumber', 'MakeFlag']].head(3)

Name ProductNumber  MakeFlag
0  Adjustable Race       AR-5381         0
1     Bearing Ball       BA-8327         0
2  BB Ball Bearing       BE-2349         1

Note that we can switch the order of `head` and the selecting of the columns, it's the exact same result, just the meaning is a bit different. In the next example, first we select the first three rows, and then we select the named columns out of those 3 rows.

In [24]:
products.head(3)[['Name', 'ProductNumber', 'MakeFlag']]

Name ProductNumber  MakeFlag
0  Adjustable Race       AR-5381         0
1     Bearing Ball       BA-8327         0
2  BB Ball Bearing       BE-2349         1

**Class Question:** What if we wanted to select a column that has a space in it? Which method from the above two would we use? Why?

---

# Selecting Row(s)

There are two ways to select rows:

* `.iloc` is position-based
* `.loc` is label-based

For now, we can only learn about `.iloc` as `.loc` relies on **Indices**, which we haven't learned about yet. So we'll come back to `.loc` later.

To select rows with `.iloc`, supply the `start` and `end` positions, as integers, with the syntax `.iloc[start:end]`. 

**Note:** `end` will not be included! (Similar to `range` which we learned about before.)

In [25]:
products.iloc[5:10]

ProductID             Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
5        317      LL Crankarm       CA-5965         0                  0   
6        318      ML Crankarm       CA-6738         0                  0   
7        319      HL Crankarm       CA-7457         0                  0   
8        320  Chainring Bolts       CB-2903         0                  0   
9        321    Chainring Nut       CN-6137         0                  0   

    Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  ...  \
5   Black               500           375           0.0        0.0  ...   
6   Black               500           375           0.0        0.0  ...   
7   Black               500           375           0.0        0.0  ...   
8  Silver              1000           750           0.0        0.0  ...   
9  Silver              1000           750           0.0        0.0  ...   

  ProductLine Class Style  ProductSubcategoryID  ProductModelID  \
5         NaN    L    NaN                   NaN             NaN   
6         NaN    M    NaN                   NaN             NaN   
7         NaN   NaN   NaN                   NaN             NaN   
8         NaN   NaN   NaN                   NaN             NaN   
9         NaN   NaN   NaN                   NaN             NaN   

         SellStartDate SellEndDate DiscontinuedDate  \
5  2008-04-30 00:00:00         NaN              NaN   
6  2008-04-30 00:00:00         NaN              NaN   
7  2008-04-30 00:00:00         NaN              NaN   
8  2008-04-30 00:00:00         NaN              NaN   
9  2008-04-30 00:00:00         NaN              NaN   

                                  rowguid                   ModifiedDate  
5  {3C9D10B7-A6B2-4774-9963-C19DCEE72FEA}  2014-02-08 10:01:36.827000000  
6  {EABB9A92-FA07-4EAB-8955-F0517B4A4CA7}  2014-02-08 10:01:36.827000000  
7  {7D3FD384-4F29-484B-86FA-4206E276FE58}  2014-02-08 10:01:36.827000000  
8  {7BE38E48-B7D6-4486-888E-F53C26735101}  2014-02-08 10:01:36.827000000  
9  {3314B1D7-EF69-4431-B6DD-DC75268BD5DF}  2014-02-08 10:01:36.827000000  

[5 rows x 25 columns]

So now combining the two concepts:

In [28]:
products[['Name', 'ProductNumber', 'MakeFlag']].iloc[5:10] #clearer

Name ProductNumber  MakeFlag
5      LL Crankarm       CA-5965         0
6      ML Crankarm       CA-6738         0
7      HL Crankarm       CA-7457         0
8  Chainring Bolts       CB-2903         0
9    Chainring Nut       CN-6137         0

You can also select columns at the same time by passing in a second integer-based range! (Although I think the previous syntax, making the column names more explicit, is clearer.)

In [26]:
products.iloc[5:10, 1:5]

Name ProductNumber  MakeFlag  FinishedGoodsFlag
5      LL Crankarm       CA-5965         0                  0
6      ML Crankarm       CA-6738         0                  0
7      HL Crankarm       CA-7457         0                  0
8  Chainring Bolts       CB-2903         0                  0
9    Chainring Nut       CN-6137         0                  0

You can read more about `.iloc` in this [Stack Overflow answer](https://stackoverflow.com/questions/31593201/how-are-iloc-ix-and-loc-different).

---

# Renaming Columns

Perhaps we want to rename our columns. There are a few options for doing this.

## Renaming Specific Columns

Here's an example of renaming **specific** columns by using a dictionary. **Note:** `inplace=False` will return a new dataframe, but leave the original dataframe untouched. Change `inplace` to `True` to modify the original dataframe.

In [31]:
products.rename(columns={'Name': 'product_name', 'ProductNumber':'product_number'}, inplace=True) #do not change until you know correct
products.head(3)

ProductID     product_name product_number  MakeFlag  FinishedGoodsFlag  \
0          1  Adjustable Race        AR-5381         0                  0   
1          2     Bearing Ball        BA-8327         0                  0   
2          3  BB Ball Bearing        BE-2349         1                  0   

  Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  ...  \
0   NaN              1000           750           0.0        0.0  ...   
1   NaN              1000           750           0.0        0.0  ...   
2   NaN               800           600           0.0        0.0  ...   

  ProductLine Class Style  ProductSubcategoryID  ProductModelID  \
0         NaN   NaN   NaN                   NaN             NaN   
1         NaN   NaN   NaN                   NaN             NaN   
2         NaN   NaN   NaN                   NaN             NaN   

         SellStartDate SellEndDate DiscontinuedDate  \
0  2008-04-30 00:00:00         NaN              NaN   
1  2008-04-30 00:00:00         NaN              NaN   
2  2008-04-30 00:00:00         NaN              NaN   

                                  rowguid                   ModifiedDate  
0  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}  2014-02-08 10:01:36.827000000  
1  {58AE3C20-4F3A-4749-A7D4-D568806CC537}  2014-02-08 10:01:36.827000000  
2  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}  2014-02-08 10:01:36.827000000  

[3 rows x 25 columns]

## Renaming All Coumns

You can also rename **ALL** columns with a new list of column names.

First let's take a look at the product columns and its type:

In [32]:
print('My columns look like:', products.columns)
print('\nAnd the type is:', type(products.columns))

My columns look like: Index(['ProductID', 'product_name', 'product_number', 'MakeFlag',
       'FinishedGoodsFlag', 'Color', 'SafetyStockLevel', 'ReorderPoint',
       'StandardCost', 'ListPrice', 'Size', 'SizeUnitMeasureCode',
       'WeightUnitMeasureCode', 'Weight', 'DaysToManufacture', 'ProductLine',
       'Class', 'Style', 'ProductSubcategoryID', 'ProductModelID',
       'SellStartDate', 'SellEndDate', 'DiscontinuedDate', 'rowguid',
       'ModifiedDate'],
      dtype='object')

And the type is: <class 'pandas.core.indexes.base.Index'>


Note that the `pd.DataFrame.columns` property can be converted (i.e. cast) to a `list` type. (Originally, it's a `pd.core.indexes.base.Index` object), by using the built-in `list()` function:

In [33]:
print('Now my columns look like:', list(products.columns))
print('\nAnd the type is:', type(list(products.columns)))

Now my columns look like: ['ProductID', 'product_name', 'product_number', 'MakeFlag', 'FinishedGoodsFlag', 'Color', 'SafetyStockLevel', 'ReorderPoint', 'StandardCost', 'ListPrice', 'Size', 'SizeUnitMeasureCode', 'WeightUnitMeasureCode', 'Weight', 'DaysToManufacture', 'ProductLine', 'Class', 'Style', 'ProductSubcategoryID', 'ProductModelID', 'SellStartDate', 'SellEndDate', 'DiscontinuedDate', 'rowguid', 'ModifiedDate']

And the type is: <class 'list'>


We can place these columns into a variable, `cols`. These strings will become the new column names:

In [34]:
cols = list(products.columns)

We can use list indexing to change the columns we want. Here, we'll rename the `ProductName` column:

In [35]:
print('Before: ', cols)

cols[1] = 'NewProductName'

print('After: ', cols)

Before:  ['ProductID', 'product_name', 'product_number', 'MakeFlag', 'FinishedGoodsFlag', 'Color', 'SafetyStockLevel', 'ReorderPoint', 'StandardCost', 'ListPrice', 'Size', 'SizeUnitMeasureCode', 'WeightUnitMeasureCode', 'Weight', 'DaysToManufacture', 'ProductLine', 'Class', 'Style', 'ProductSubcategoryID', 'ProductModelID', 'SellStartDate', 'SellEndDate', 'DiscontinuedDate', 'rowguid', 'ModifiedDate']
After:  ['ProductID', 'NewProductName', 'product_number', 'MakeFlag', 'FinishedGoodsFlag', 'Color', 'SafetyStockLevel', 'ReorderPoint', 'StandardCost', 'ListPrice', 'Size', 'SizeUnitMeasureCode', 'WeightUnitMeasureCode', 'Weight', 'DaysToManufacture', 'ProductLine', 'Class', 'Style', 'ProductSubcategoryID', 'ProductModelID', 'SellStartDate', 'SellEndDate', 'DiscontinuedDate', 'rowguid', 'ModifiedDate']


In [36]:
products.columns = cols

print(products.columns)

Index(['ProductID', 'NewProductName', 'product_number', 'MakeFlag',
       'FinishedGoodsFlag', 'Color', 'SafetyStockLevel', 'ReorderPoint',
       'StandardCost', 'ListPrice', 'Size', 'SizeUnitMeasureCode',
       'WeightUnitMeasureCode', 'Weight', 'DaysToManufacture', 'ProductLine',
       'Class', 'Style', 'ProductSubcategoryID', 'ProductModelID',
       'SellStartDate', 'SellEndDate', 'DiscontinuedDate', 'rowguid',
       'ModifiedDate'],
      dtype='object')


Note that our first column name has changed from `ProductName` to `NewProductName`.

Finally, we can set the `pd.DataFrame.columns` instance variable, to the new `cols` list, overwriting the existing columns header names:

In [37]:
products.columns = cols
print(products.columns)

Index(['ProductID', 'NewProductName', 'product_number', 'MakeFlag',
       'FinishedGoodsFlag', 'Color', 'SafetyStockLevel', 'ReorderPoint',
       'StandardCost', 'ListPrice', 'Size', 'SizeUnitMeasureCode',
       'WeightUnitMeasureCode', 'Weight', 'DaysToManufacture', 'ProductLine',
       'Class', 'Style', 'ProductSubcategoryID', 'ProductModelID',
       'SellStartDate', 'SellEndDate', 'DiscontinuedDate', 'rowguid',
       'ModifiedDate'],
      dtype='object')


Now we can display the final result of the changing the `ProductName` column to `NewName`:

In [38]:
products.head(3)

ProductID   NewProductName product_number  MakeFlag  FinishedGoodsFlag  \
0          1  Adjustable Race        AR-5381         0                  0   
1          2     Bearing Ball        BA-8327         0                  0   
2          3  BB Ball Bearing        BE-2349         1                  0   

  Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  ...  \
0   NaN              1000           750           0.0        0.0  ...   
1   NaN              1000           750           0.0        0.0  ...   
2   NaN               800           600           0.0        0.0  ...   

  ProductLine Class Style  ProductSubcategoryID  ProductModelID  \
0         NaN   NaN   NaN                   NaN             NaN   
1         NaN   NaN   NaN                   NaN             NaN   
2         NaN   NaN   NaN                   NaN             NaN   

         SellStartDate SellEndDate DiscontinuedDate  \
0  2008-04-30 00:00:00         NaN              NaN   
1  2008-04-30 00:00:00         NaN              NaN   
2  2008-04-30 00:00:00         NaN              NaN   

                                  rowguid                   ModifiedDate  
0  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}  2014-02-08 10:01:36.827000000  
1  {58AE3C20-4F3A-4749-A7D4-D568806CC537}  2014-02-08 10:01:36.827000000  
2  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}  2014-02-08 10:01:36.827000000  

[3 rows x 25 columns]

---

# Common Column Operations

While this is non-comprehensive, these are a few key column-specific data checks.

## Descriptive Statistics:

The mean, standard deviation, minimum, first quartile, median, third quartile, and maximum of a bunch of numbers.

* **Mean:** The average of the column
* **[Standard Deviation](https://en.wikipedia.org/wiki/Standard_deviation):** From statistics -- measures the variation around the mean.
* **Min:** The smallest value in the column
* **Max:** The largest value in the column
* **Quartile:** A quartile is one fourth of our data
   * **First quartile:** This is the bottom most 25 percent
   * **Median:** The middle value. (Line all values biggest to smallest - median is the middle!) Also the 50th percentile
   * **Third quartile:** This the the top 75 percentile of our data
   
![](assets/quartiles.svg)

In Pandas, we can get these statistics with `.describe()` ([docs](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html)).

**Note:** `.describe()` by default operates on numeric datatypes, but it can also operate on strings and timestamps.

Here, we use `.describe()` on the `MakeFlag`, `SafetyStockLevel`, and `StandardCost` fields.

In [39]:
products[['SafetyStockLevel', 'StandardCost', 'ListPrice']].describe()

SafetyStockLevel  StandardCost    ListPrice
count        504.000000    504.000000   504.000000
mean         535.150794    258.602961   438.666250
std          374.112954    461.632808   773.602843
min            4.000000      0.000000     0.000000
25%          100.000000      0.000000     0.000000
50%          500.000000     23.372200    49.990000
75%         1000.000000    317.075825   564.990000
max         1000.000000   2171.294200  3578.270000

## Value Counts

`pd.Series.value_counts()` count the occurrence of each value within a Series.

The following shows the most popular product colors (aggregated by count, descending by default):

In [41]:
products['Color'].value_counts()

Black           93
Silver          43
Red             38
Yellow          36
Blue            26
Multi            8
Silver/Black     7
White            4
Grey             1
Name: Color, dtype: int64

## Unique values

`pd.Series.unique()` can determine the number of distinct values within a given Series.

What are the unique colors for the products?

In [42]:
products['Color'].unique()

array([nan, 'Black', 'Silver', 'Red', 'White', 'Blue', 'Multi', 'Yellow',
       'Grey', 'Silver/Black'], dtype=object)

How many distinct colors are there? (With `nunique()`)

In [43]:
products['Color'].nunique()

9

Notice that `nunique()` doesn't count nulls by default. We can also include nulls with the parameter `dropna=False`:

In [44]:
products['Color'].nunique(dropna=False)

10

---

# Filtering

## Filtering on a Single Condition

Filtering and sorting are key processes that allow us to drill into the 'nitty gritty' and cross sections of our dataset.

To filter, we use a process called **Boolean Filtering**, wherein we define a Boolean condition, and use that Boolean condition to filer on our DataFrame.

Recall: our given dataset has a column `Color`. Let's see if we can find all products that are `Black`. Let's take a look at the first 10 rows by Color to see how it looks as-is:

In [45]:
products['Color'].head(10)

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
5     Black
6     Black
7     Black
8    Silver
9    Silver
Name: Color, dtype: object

By applying a **boolean mask** to this dataframe, `== 'Black'`, we can get the following **filter**:

In [46]:
products['Color'].head(10) == 'Black' #a filter

0    False
1    False
2    False
3    False
4    False
5     True
6     True
7     True
8    False
9    False
Name: Color, dtype: bool

Now we can apply the **filter** to our full dataframe. We want the rows where the filter is `True`, and we do'nt want the rows where the filter is `False`.

The result is a dataframe that only has rows where `Color` is `Black`:

In [47]:
products[  products['Color'] == 'Black' ].head(5)

ProductID             NewProductName product_number  MakeFlag  \
5          317                LL Crankarm        CA-5965         0   
6          318                ML Crankarm        CA-6738         0   
7          319                HL Crankarm        CA-7457         0   
10         322                  Chainring        CR-7833         0   
209        680  HL Road Frame - Black, 58     FR-R92B-58         1   

     FinishedGoodsFlag  Color  SafetyStockLevel  ReorderPoint  StandardCost  \
5                    0  Black               500           375          0.00   
6                    0  Black               500           375          0.00   
7                    0  Black               500           375          0.00   
10                   0  Black              1000           750          0.00   
209                  1  Black               500           375       1059.31   

     ListPrice  ... ProductLine Class Style  ProductSubcategoryID  \
5          0.0  ...         NaN    L    NaN                   NaN   
6          0.0  ...         NaN    M    NaN                   NaN   
7          0.0  ...         NaN   NaN   NaN                   NaN   
10         0.0  ...         NaN   NaN   NaN                   NaN   
209     1431.5  ...          R     H     U                   14.0   

     ProductModelID        SellStartDate SellEndDate DiscontinuedDate  \
5               NaN  2008-04-30 00:00:00         NaN              NaN   
6               NaN  2008-04-30 00:00:00         NaN              NaN   
7               NaN  2008-04-30 00:00:00         NaN              NaN   
10              NaN  2008-04-30 00:00:00         NaN              NaN   
209             6.0  2008-04-30 00:00:00         NaN              NaN   

                                    rowguid                   ModifiedDate  
5    {3C9D10B7-A6B2-4774-9963-C19DCEE72FEA}  2014-02-08 10:01:36.827000000  
6    {EABB9A92-FA07-4EAB-8955-F0517B4A4CA7}  2014-02-08 10:01:36.827000000  
7    {7D3FD384-4F29-484B-86FA-4206E276FE58}  2014-02-08 10:01:36.827000000  
10   {F0AC2C4D-1A1F-4E3C-B4D9-68AEA0EC1CE4}  2014-02-08 10:01:36.827000000  
209  {43DD68D6-14A4-461F-9069-55309D90EA7E}  2014-02-08 10:01:36.827000000  

[5 rows x 25 columns]

If you want, you can make a copy of the black products with the `copy()` method. Then you'll have an independent Data Frame for just black products.

In [48]:
black_products = products[products['Color'] == 'Black'].copy()
black_products.shape

(93, 25)

## Filtering on a Single Condition -- Example 2

OK, let's try something else. Let's calculate the **average ListPrice** for **salable products**.

Think: What are the component parts of this problem?

* First, we need to get salable items. Use your data dictionary from the beginning of this lesson... salable products are "Finished"!

In [50]:
products[products['FinishedGoodsFlag'] == 1].head(3)

ProductID             NewProductName product_number  MakeFlag  \
209        680  HL Road Frame - Black, 58     FR-R92B-58         1   
210        706    HL Road Frame - Red, 58     FR-R92R-58         1   
211        707      Sport-100 Helmet, Red      HL-U509-R         0   

     FinishedGoodsFlag  Color  SafetyStockLevel  ReorderPoint  StandardCost  \
209                  1  Black               500           375     1059.3100   
210                  1    Red               500           375     1059.3100   
211                  1    Red                 4             3       13.0863   

     ListPrice  ... ProductLine Class Style  ProductSubcategoryID  \
209    1431.50  ...          R     H     U                   14.0   
210    1431.50  ...          R     H     U                   14.0   
211      34.99  ...          S    NaN   NaN                  31.0   

     ProductModelID        SellStartDate SellEndDate DiscontinuedDate  \
209             6.0  2008-04-30 00:00:00         NaN              NaN   
210             6.0  2008-04-30 00:00:00         NaN              NaN   
211            33.0  2011-05-31 00:00:00         NaN              NaN   

                                    rowguid                   ModifiedDate  
209  {43DD68D6-14A4-461F-9069-55309D90EA7E}  2014-02-08 10:01:36.827000000  
210  {9540FF17-2712-4C90-A3D1-8CE5568B2462}  2014-02-08 10:01:36.827000000  
211  {2E1EF41A-C08A-4FF6-8ADA-BDE58B64A712}  2014-02-08 10:01:36.827000000  

[3 rows x 25 columns]

* Now, we need to find average list price of those above items. Let's just get the 'ListPrice' column for starters.

In [52]:
products[products['FinishedGoodsFlag'] == 1]['ListPrice'].head(3)

209    1431.50
210    1431.50
211      34.99
Name: ListPrice, dtype: float64

To get the average of that column, just take `.mean()`

In [54]:
products[products['FinishedGoodsFlag'] == 1]['ListPrice'].mean()

744.595220338982

We can take a shortcut and just use `.describe()` here:

In [55]:
products[products['FinishedGoodsFlag'] == 1]['ListPrice'].describe()
#list_prices = products[products['FinishedGoodsFlag'] == 1]['ListPrice']

#list_prices.describe()

count     295.000000
mean      744.595220
std       892.563172
min         2.290000
25%        66.745000
50%       337.220000
75%      1100.240000
max      3578.270000
Name: ListPrice, dtype: float64

**Sneak peek**: Another handy trick is to use `.hist()` to get a distribution of a continuous variable - in this case, `ListPrice`. We'll cover this more in future lessons:

count     295.000000
mean      744.595220
std       892.563172
min         2.290000
25%        66.745000
50%       337.220000
75%      1100.240000
max      3578.270000
Name: ListPrice, dtype: float64

## Filtering on Multiple Conditions

Here, we will filter on _multiple conditions_. Before, we filtered on rows where `Color` was `Black`. We also filtered where `FinishedGoodsFlag` was equal to `1`.

Let's see what happens when we filter on *both* simultaneously. 

The syntax for multiple conditions is:

`df[ (df['col1'] == value1) & (df['col2'] == value2) ]`

Or, more simply:

`df[ (CONDITION 1) & (CONDITION 2) ]`

Which eventually may, on a row-by-row basis, evaluate to something like:

`df[ True & False ]`

If the end result is `True`, the row is included. If the end result is `False`, the row is omitted.

_Don't forget parentheses in your conditions!_ This is a common mistake.

Let's look at a table where `Color` is `Black`, **AND** `FinishedGoodsFlag` is `1`

In [58]:
products[(products['Color'] == 'Black') & (products['FinishedGoodsFlag'] == 1)].head(3)

ProductID             NewProductName product_number  MakeFlag  \
209        680  HL Road Frame - Black, 58     FR-R92B-58         1   
212        708    Sport-100 Helmet, Black        HL-U509         0   
226        722  LL Road Frame - Black, 58     FR-R38B-58         1   

     FinishedGoodsFlag  Color  SafetyStockLevel  ReorderPoint  StandardCost  \
209                  1  Black               500           375     1059.3100   
212                  1  Black                 4             3       13.0863   
226                  1  Black               500           375      204.6251   

     ListPrice  ... ProductLine Class Style  ProductSubcategoryID  \
209    1431.50  ...          R     H     U                   14.0   
212      34.99  ...          S    NaN   NaN                  31.0   
226     337.22  ...          R     L     U                   14.0   

     ProductModelID        SellStartDate SellEndDate DiscontinuedDate  \
209             6.0  2008-04-30 00:00:00         NaN              NaN   
212            33.0  2011-05-31 00:00:00         NaN              NaN   
226             9.0  2011-05-31 00:00:00         NaN              NaN   

                                    rowguid                   ModifiedDate  
209  {43DD68D6-14A4-461F-9069-55309D90EA7E}  2014-02-08 10:01:36.827000000  
212  {A25A44FB-C2DE-4268-958F-110B8D7621E2}  2014-02-08 10:01:36.827000000  
226  {2140F256-F705-4D67-975D-32DE03265838}  2014-02-08 10:01:36.827000000  

[3 rows x 25 columns]

Next we have an example of a `ListPrice` greater than 50, **OR** a product size that is not equal to `XL`:

In [59]:
products[(products['ListPrice'] > 50) | (products['Size'] != 'XL')].head(3)

ProductID   NewProductName product_number  MakeFlag  FinishedGoodsFlag  \
0          1  Adjustable Race        AR-5381         0                  0   
1          2     Bearing Ball        BA-8327         0                  0   
2          3  BB Ball Bearing        BE-2349         1                  0   

  Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  ...  \
0   NaN              1000           750           0.0        0.0  ...   
1   NaN              1000           750           0.0        0.0  ...   
2   NaN               800           600           0.0        0.0  ...   

  ProductLine Class Style  ProductSubcategoryID  ProductModelID  \
0         NaN   NaN   NaN                   NaN             NaN   
1         NaN   NaN   NaN                   NaN             NaN   
2         NaN   NaN   NaN                   NaN             NaN   

         SellStartDate SellEndDate DiscontinuedDate  \
0  2008-04-30 00:00:00         NaN              NaN   
1  2008-04-30 00:00:00         NaN              NaN   
2  2008-04-30 00:00:00         NaN              NaN   

                                  rowguid                   ModifiedDate  
0  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}  2014-02-08 10:01:36.827000000  
1  {58AE3C20-4F3A-4749-A7D4-D568806CC537}  2014-02-08 10:01:36.827000000  
2  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}  2014-02-08 10:01:36.827000000  

[3 rows x 25 columns]

---

# Sorting

We can sort our DataFrame as well.

Let's sort by `StandardCost` on a descending basis:

In [62]:
products.sort_values('StandardCost', ascending=False).head(3)

ProductID    NewProductName product_number  MakeFlag  FinishedGoodsFlag  \
253        749  Road-150 Red, 62     BK-R93R-62         1                  1   
254        750  Road-150 Red, 44     BK-R93R-44         1                  1   
255        751  Road-150 Red, 48     BK-R93R-48         1                  1   

    Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  ...  \
253   Red               100            75     2171.2942    3578.27  ...   
254   Red               100            75     2171.2942    3578.27  ...   
255   Red               100            75     2171.2942    3578.27  ...   

    ProductLine Class Style  ProductSubcategoryID  ProductModelID  \
253          R     H     U                    2.0            25.0   
254          R     H     U                    2.0            25.0   
255          R     H     U                    2.0            25.0   

           SellStartDate          SellEndDate DiscontinuedDate  \
253  2011-05-31 00:00:00  2012-05-29 00:00:00              NaN   
254  2011-05-31 00:00:00  2012-05-29 00:00:00              NaN   
255  2011-05-31 00:00:00  2012-05-29 00:00:00              NaN   

                                    rowguid                   ModifiedDate  
253  {BC621E1F-2553-4FDC-B22E-5E44A9003569}  2014-02-08 10:01:36.827000000  
254  {C19E1136-5DA4-4B40-8758-54A85D7EA494}  2014-02-08 10:01:36.827000000  
255  {D10B7CC1-455E-435B-A08F-EC5B1C5776E9}  2014-02-08 10:01:36.827000000  

[3 rows x 25 columns]

It's also possible to sort by multiple columns, but passing in a list of the columns to sort ([docs](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html)).

The next one is a little more advanced, but it demonstrates a few things:

* Conversion of a `numpy.ndarray` object (return type of `pd.Series.unique()`) into a `pd.Series` object
* `pd.Series.sort_values()` with the `by=` kwarg omitted (if only one column exists, `by=` doesn't need to specified)
* Alphabetical sort of a string field, `ascending=True` means A->Z

In [67]:
pd.Series(products['Color'].unique()).sort_values(ascending=True)

1           Black
5            Blue
8            Grey
6           Multi
3             Red
2          Silver
9    Silver/Black
4           White
7          Yellow
0             NaN
dtype: object

---

# A Detour: The Pandas Index

In Pandas, an [index](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Index.html) is an "immutable `ndarray` implementing an ordered, sliceable set" (not to be confused with *list indexes*, which you just saw again).

It is the basic object storing axis labels for all pandas objects. 

Think of it as a 'row address' for your table/dataframe. It is best practice to explicitly set the index of your dataframe, as indices are commonly used as [primary keys](https://en.wikipedia.org/wiki/Primary_key) which can be used to [join](https://www.w3schools.com/sql/sql_join.asp) your dataframe to other dataframes.

The dataframe can store different types (ints, string, datetime), and when data is imported, Pandas will automatically assign a number to each row, starting at zero and counting up. You can overwrite this, which is what we are going to do.

First, let's take a look again at the layout of the dataframe. Notice the leftmost column is just an ascending number from 0. This is the auto-generated index that was created upon import:

More information about the index:

Note that our auto-generated index has no name:

Here we are looking at three columns;
the one on the left is the index (automatically generated upon import by pandas)
'ProductID' is our PK (primary key) from our imported table. 'Name' is a data column.

Notice that the generated index starts at zero and our PK starts at 1.

Setting the index overwrites the automatically generated index with our PK, which resided in the `ProductID` column:

Note how our index property has changed as a result:

And our index name has also inherited the name of our `ProductID` column:

And now if we example the Dataframe, the leftmost column looks different!

---

# We Do: Pandas Exercises

Let's do these Exercises. Don't hesitate to look at code we wrote together!

## Exercise 1:

Print the first 4 rows of the whole DataFrame

## Exercise 2:

How many rows are in the dataframe? Return the answer as an int.

## Exercise 3:

How many columns? Return the answer as an int.

## Exercise 4:

How many different product lines are there?

## Exercise 5:

What are the values of these product lines?

## Exercise 6:

Do the number of values for the product lines match the number you have using `.nunique()`? Why or why not?

--- Type your answer in this Markdown cell ---

## Exercise 7:

Take the output from your previous answer (using `.unique()`). Select the label corresponding to the `Road` product line using list indexing notation. How many characters are in this string?

Do you notice anything odd about this?

There is a space after R!!!!! The horror!!!!!

## Exercise 8:

How many products are there for the `Road` product line? Don't forget what you just saw above! Return your answer as an int.

## Exercise 9:

How many products are there in the `Women's` `Mountain` category? Return your answer as an int. _Hint: Use the data dictionary above!_

## Exercise 10:

**Very Hard Challenge:** What are the top 3 _most expensive list price_ product that are either in the `Women's` `Mountain` category, _OR_ `Silver` in `Color`? Return your answer as a DataFrame object, with the `ProductID` index, `NewName` relabeled as `Name`, and `ListPrice` columns. Perform the statement in one execution, and do not mutate the source DataFrame.

---

# Summary

We covered a lot of ground! It's ok if this takes a while to gel, learning Pandas can be frustrating!

It's common to refer back to your own code *all the time.* Don't hesistate to reference this notebook! 🐼